vm/systemSmartContracts/delegation_test.go

```diff
@@ -4246,3 +4246,455 @@ func TestDelegation_GetDelegationManagementShouldWork(t *testing.T) {
+		&mock.ArgumentParserMock{},
+		&mock.AccountsStub{},
+		&mock.RaterMock{},
+	)
+	_ = eei.SetSystemSCContainer(&mock.SystemSCContainerStub{GetCalled: func(key []byte) (vm.SystemSmartContract, error) {
+		return &mock.SystemSCStub{ExecuteCalled: func(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+			return vmcommon.Ok
+		}}, nil
+	}})
+
+	args.Eei = eei
+	args.DelegationSCConfig.MaxServiceFee = 10000
+	args.DelegationSCConfig.MinServiceFee = 0
+	d, _ := NewDelegationSystemSC(args)
+
+	addr := []byte("address")
+	_, returnCode := d.getAndVerifyValidatorData(addr)
+	assert.Equal(t, eei.returnMessage, vm.ErrEmptyStorage.Error())
+	assert.Equal(t, returnCode, vmcommon.UserError)
+
+	eei.SetStorageForAddress(d.validatorSCAddr, addr, addr)
+	_, returnCode = d.getAndVerifyValidatorData(addr)
+	assert.Equal(t, returnCode, vmcommon.UserError)
+
+	validatorData := &ValidatorDataV2{
+		RewardAddress:   []byte("randomAddress"),
+		TotalSlashed:    big.NewInt(0),
+		TotalUnstaked:   big.NewInt(0),
+		TotalStakeValue: big.NewInt(0),
+		UnstakedInfo:    []*UnstakedValue{{UnstakedValue: big.NewInt(10)}},
+		NumRegistered:   3,
+		BlsPubKeys:      [][]byte{[]byte("firsstKey"), []byte("secondKey"), []byte("thirddKey")},
+	}
+	marshaledData, _ := d.marshalizer.Marshal(validatorData)
+	eei.SetStorageForAddress(d.validatorSCAddr, addr, marshaledData)
+
+	eei.returnMessage = ""
+	_, returnCode = d.getAndVerifyValidatorData(addr)
+	assert.Equal(t, returnCode, vmcommon.UserError)
+	assert.Equal(t, eei.returnMessage, "invalid reward address on validator data")
+
+	validatorData.RewardAddress = addr
+	marshaledData, _ = d.marshalizer.Marshal(validatorData)
+	eei.SetStorageForAddress(d.validatorSCAddr, addr, marshaledData)
+
+	eei.returnMessage = ""
+	_, returnCode = d.getAndVerifyValidatorData(addr)
+	assert.Equal(t, returnCode, vmcommon.UserError)
+
+	managementData := &DelegationManagement{
+		NumOfContracts:      0,
+		LastAddress:         vm.FirstDelegationSCAddress,
+		MinServiceFee:       0,
+		MaxServiceFee:       100,
+		MinDeposit:          big.NewInt(100),
+		MinDelegationAmount: big.NewInt(100),
+	}
+	marshaledData, _ = d.marshalizer.Marshal(managementData)
+	eei.SetStorageForAddress(d.delegationMgrSCAddress, []byte(delegationManagementKey), marshaledData)
+
+	eei.returnMessage = ""
+	_, returnCode = d.getAndVerifyValidatorData(addr)
+	assert.Equal(t, returnCode, vmcommon.UserError)
+	assert.Equal(t, eei.returnMessage, "not enough stake to make delegation contract")
+
+	validatorData.TotalStakeValue.SetUint64(10000)
+	marshaledData, _ = d.marshalizer.Marshal(validatorData)
+	eei.SetStorageForAddress(d.validatorSCAddr, addr, marshaledData)
+
+	eei.returnMessage = ""
+	_, returnCode = d.getAndVerifyValidatorData(addr)
+	assert.Equal(t, returnCode, vmcommon.UserError)
+	assert.Equal(t, eei.returnMessage, "clean unStaked info before changing validator to delegation contract")
+}
+
+func TestDelegation_initFromValidatorData(t *testing.T) {
```

Comment:
This is the longest unit test I ever seen :)

Reply:
A lot of cases are tested here